src/middleware/withDelegationsStorage.js

```diff
@@ -0,0 +1,95 @@
+import { Delegation } from '@ucanto/core'
+import { ok, error, Failure } from '@ucanto/server'
+
+/**
+ * @import * as Ucanto from '@ucanto/interface'
+ * @import { Middleware } from '@web3-storage/gateway-lib'
+ * @typedef {import('./withDelegationsStorage.types.js').DelegationsStorageContext} DelegationsStorageContext
+ * @typedef {import('./withDelegationsStorage.types.js').DelegationsStorageEnvironment} DelegationsStorageEnvironment
+ */
+export class InvalidDelegation extends Failure {
+  get name() {
+    return /** @type {const} */ ('InvalidDelegation')
+  }
+}
+
+export class DelegationNotFound extends Failure {
+  get name() {
+    return /** @type {const} */ ('DelegationNotFound')
+  }
+}
+
+export class StoreOperationFailed extends Failure {
+  get name() {
+    return /** @type {const} */ ('StoreOperationFailed')
+  }
+}
```

Comment:
You need to implement a `describe` method in these, unfortunately the way `Failure` is implemented causes a recursive call loop:

```js
import { Failure } from '@ucanto/server'

class DelegationNotFound extends Failure {
  get name() {
    return /** @type {const} */ ('DelegationNotFound')
  }
}

const err = new DelegationNotFound()
console.log(err.message)
```

```
RangeError: Maximum call stack size exceeded
    at DelegationNotFound.toString (<anonymous>)
    at DelegationNotFound.describe (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:65:17)
    at get message (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:68:17)
    at DelegationNotFound.toString (<anonymous>)
    at DelegationNotFound.describe (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:65:17)
    at get message (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:68:17)
    at DelegationNotFound.toString (<anonymous>)
    at DelegationNotFound.describe (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:65:17)
    at get message (file:///Users/alan/Code/storacha/w3up/node_modules/.pnpm/@ucanto+core@10.0.1/node_modules/@ucanto/core/src/result.js:68:17)
    at DelegationNotFound.toString (<anonymous>)
```

Something like this:


```js
class DelegationNotFound extends Failure {
  static name = /** @type {const} */ ('DelegationNotFound')
  #reason

  /** @param {string} [reason] */
  constructor (reason) {
    super()
    this.#reason = reason
  }

  get name() {
    return DelegationNotFound.name
  }

  describe() {
    return this.#reason ?? 'Delegation not found'
  }
}
```

Note: optional, but the static `name` property I added here allows users to switch on the error name without using magic string:

```js
if (res.error?.name === DelegationNotFound.name) {
  // do something with this specific error
}
```